pkg/tuf/options.go

```diff
@@ -0,0 +1,135 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tuf
+
+import (
+	"embed"
+	"os"
+	"path/filepath"
+
+	"github.com/theupdateframework/go-tuf/v2/metadata/fetcher"
+)
+
+//go:embed repository
+var embeddedRepo embed.FS
+
+const DefaultMirror = "https://tuf-repo-cdn.sigstore.dev"
+
+// Options represent the various options for a Sigstore TUF Client
+type Options struct {
+	// CacheValidity period in days (default 1)
+	CacheValidity int
+	// ForceCache controls if the cache should be used without update
+	// as long as the metadata is valid
+	ForceCache bool
+	// Root is the TUF trust anchor
+	Root []byte
+	// CachePath is the location on disk for TUF cache
+	// (default $HOME/.sigstore/tuf)
+	CachePath string
+	// RepositoryBaseURL is the TUF repository location URL
+	// (default https://tuf-repo-cdn.sigstore.dev)
+	RepositoryBaseURL string
+	// DisableLocalCache mode allows a client to work on a read-only
+	// files system if this is set, cache path is ignored.
+	DisableLocalCache bool
+	// DisableConsistentSnapshot
+	DisableConsistentSnapshot bool
+	// Fetcher is the metadata fetcher
+	Fetcher fetcher.Fetcher
+}
+
+// WithCacheValidity sets the cache validity period in days
+func (o *Options) WithCacheValidity(days int) *Options {
+	o.CacheValidity = days
+	return o
+}
+
+// WithForceCache forces the client to use the cache without updating
+func (o *Options) WithForceCache() *Options {
+	o.ForceCache = true
+	return o
+}
+
+// WithRoot sets the TUF trust anchor
+func (o *Options) WithRoot(root []byte) *Options {
+	o.Root = root
+	return o
+}
+
+// WithCachePath sets the location on disk for TUF cache
+func (o *Options) WithCachePath(path string) *Options {
+	o.CachePath = path
+	return o
+}
+
+// WithRepositoryBaseURL sets the TUF repository location URL
+func (o *Options) WithRepositoryBaseURL(url string) *Options {
+	o.RepositoryBaseURL = url
+	return o
+}
+
+// WithDisableLocalCache sets the client to work on a read-only file system
+func (o *Options) WithDisableLocalCache() *Options {
+	o.DisableLocalCache = true
+	return o
+}
+
+// WithDisableConsistentSnapshot sets the client to disable consistent snapshot
+func (o *Options) WithDisableConsistentSnapshot() *Options {
+	o.DisableConsistentSnapshot = true
+	return o
+}
+
+// WithFetcher sets the metadata fetcher
+func (o *Options) WithFetcher(f fetcher.Fetcher) *Options {
+	o.Fetcher = f
+	return o
+}
+
+// DefaultOptions returns an options struct for the public good instance
+func DefaultOptions() *Options {
+	var opts Options
+	var err error
+
+	opts.Root = DefaultRoot()
+	home, err := os.UserHomeDir()
+	if err != nil {
+		// Fall back to using a TUF repository in the temp location
+		home = os.TempDir()
+	}
+	opts.CacheValidity = 1
```

Comment:
Should we have a cache by default? I think cache is expected to be used in the case of airgapped environments, so should we have this be 0 and just rely on timestamp expiration for when to fetch metadata by default?

Reply:
Back in the summer of 2022 when we where discussing this, 1 day was proposed https://docs.google.com/document/d/1QWBvpwYxOy9njAmd8vpizNQpPti9rd5ugVhji0r3T4c/edit#heading=h.s9pcbj5k0o1z

In retrospect, I think that the tuf package maybe should set the cache validity to 0 (i.e no cache) but each client should set the value to the best in their domain. This way we minimize any possible footguns (i.e secure by default). Even if a single is quick, I could still see possible issues being opened when the client (e.g. cosign) does not immediately get the new root. 

I can change this. cc @codysoyland

Reply:
How does this interact with the target cache files? I don’t believe it does, correct?

Reply:
Reading this comment again:

> Should we have a cache by default? I think cache is expected to be used in the case of airgapped environments, so should we have this be 0 and just rely on timestamp expiration for when to fetch metadata by default?

This is not really how it works. If caching is not used, the client will always try to get a new timestamp and root, regardless if the local timestamp is expired or not (this is according to the TUF spec). 

The cache allows us to control if this (full TUF online refresh) should be performed on every invocation, or only once per day. But I'm still thinking that maybe having the cache off by default is a better option, as a full TUF update is not that expensive, if the client is up to date, it's two egress HTTP requests (one for root and one for timestamp). So the cache is not really with arigapped, it's more a way to speed up invocations where e.g. frequent cosign verify is performed in a short period of time.

Reply:
> How does this interact with the target cache files? I don’t believe it does, correct? As in, the client will still cache targets on disk regardless of this configuration.

Exactly, it's only the metadata that is cached in non-standard way to avoid a full metadata refresh against the remote TUF repository.

Reply:
Since we're using the `UnsafeLocalMode` updater for the initial update, won't cache validity only matter if the timestamp is expired?

Reply:
Also, should cache validity be checked in `GetTarget` too? Otherwise `GetTargetInfo` forces a refresh if metadata is expired.

Reply:
I might be wrong about the above, so maybe I should ask that as a question: In what cases is the underlying tuf client implicitly refreshing data?

Reply:
> Since we're using the `UnsafeLocalMode` updater for the initial update, won't cache validity only matter if the timestamp is expired?

No, timestamp expiration is always respected. As the timestamp may expire at any time, or a new root (or target) can be published, so the client always looks for that on every invocation. This is the behaviour that's controlled via `CacheValidity`

> Also, should cache validity be checked in GetTarget too? Otherwise GetTargetInfo forces a refresh if metadata is expired.

Yes, if the metadata is expired, this will happen. If this is in an airgapped environment this would fail, but for that, I would think a better approach is to just provide the `trusted_root.json` instead of using TUF, as we won't get any real security benefit from it. And also, I would think that the normal scenario is that a client does:
1. Create a TUF client
2. Get `trusted_root.json` target
and the two calls happens during initialization, pretty close to each other in time. And once the `trusted_root.json` is retrieved, it's used for the rest of the program's execution?

Reply:
Ah, I took another pass over the code and now see what I was missing. In the current Sigstore TUF client, we have it set up where if the timestamp is not expired, no additional metadata is fetched. With the current implementation, if you wanted to replicate that behavior, you would set `CacheValidity` to be greater than the timestamp expiration - Then it would not refresh until the timestamp expires, which will then force a refresh.

Have you considered having this behavior controlled via a boolean rather than a time period? The current implementation does give finer control though. One idea to improve UX, maybe a NO_CACHE and MAX_CACHE constant, set to 0 and MAX_INT respectively? Then when setting up the config, you can choose to a) follow the TUF spec with NO_CACHE, b) only refresh when absolutely necessary with MAX_CACHE, or c) fine-tune based on needs.

I might also add a comment making this explicit that `CacheValidity` is not for airgapped environments, since it still does respect timestamp expiration.

Reply:
> Have you considered having this behavior controlled via a boolean rather than a time period? 

I can't speak for @kommendorkapten, but in my opinion, the `CacheValidity` flag works well for systems that perform a lot of verifications, without having to manually bump the cache daily, and avoids the risk of missing new metadata if the repo's timestamp interval is long. However, I could see the argument of changing it to hours or time.Duration, or adding a few constants as you suggested.

> I might also add a comment making this explicit that CacheValidity is not for airgapped environments, since it still does respect timestamp expiration.

What is your opinion on the correct behavior for airgapped environments? Would you suggest having an option to disregard an expired timestamp? That is certainly possible, but I would probably want to slap the `Insecure` prefix in front of that option. 😆

Reply:
> CacheValidity flag works well for systems that perform a lot of verifications, without having to manually bump the cache daily, and avoids the risk of missing new metadata if the repo's timestamp interval is long

This sounds good. I think the constants are a nice compromise for those who are configuring the client and want a simple "boolean" approach of "no cache" vs "cache up to timestamp expiration".

> What is your opinion on the correct behavior for airgapped environments? Would you suggest having an option to disregard an expired timestamp?

I think we should omit airgapped support for now, there's much more nuance with getting this right, given it doesn't follow the TUF spec. I agree with @kommendorkapten that for airgapped clients, a reasonable approach is to fetch the trust root out of band. The python client has been thinking about what it means to be [truly offline](https://github.com/sigstore/sigstore-python/issues/483) too.

Reply:
Got it, I missed @kommendorkapten's comment on that, but I agree with you both!